src/server/pfs/s3/bucket.go

```diff
@@ -43,50 +43,62 @@ type bucketController struct {
 	logger *logrus.Entry
 }
 
-func (c bucketController) GetLocation(r *http.Request, bucket string, result *s2.LocationConstraint) error {
+func newBucketController(pc *client.APIClient, logger *logrus.Entry) *bucketController {
+	c := bucketController{
+		pc:     pc,
+		logger: logger,
+	}
+
+	return &c
+}
+
+func (c *bucketController) GetLocation(r *http.Request, bucket string) (location string, err error) {
 	repo, branch, err := bucketArgs(r, bucket)
 	if err != nil {
-		return err
+		return
 	}
 
 	_, err = c.pc.InspectBranch(repo, branch)
 	if err != nil {
-		return maybeNotFoundError(r, err)
+		err = maybeNotFoundError(r, err)
+		return
 	}
 
-	result.Location = "PACHYDERM"
-	return nil
+	location = globalLocation
+	return
 }
 
-func (c bucketController) List(r *http.Request, bucket string, result *s2.ListBucketResult) error {
+func (c *bucketController) ListObjects(r *http.Request, bucket, prefix, marker, delimiter string, maxKeys int) (contents []s2.Contents, commonPrefixes []s2.CommonPrefixes, isTruncated bool, err error) {
 	repo, branch, err := bucketArgs(r, bucket)
 	if err != nil {
-		return err
+		return
```

Comment:
Using named return values is generally frowned upon because it's easy to shoot yourself in the foot with them, I've done it several times. The only time we use them is if we need to return errors from deferred functions. Here's an example of how things can go wrong:

```golang
package main

import (
	"fmt"
)

func foo() (err error) {
	if true {
		x, err := 1, fmt.Errorf("error")
		fmt.Printf("x: %d\n", x)
		fmt.Printf("err: %v\n", err)
	}
	return
}

func main() {
	fmt.Println(foo())
}
```

It seems like you're returning the error but you're actually not because of scoping rules. Afaict everything you're doing here is fine, but I had to spend a while staring at it to convince myself that there weren't any scoping issues like this, so I think it'd be clearer if you just returned errors normally.

Reply:
After continuing to build on these new interface definitions, I very much agree. But given all the stuff in-flight that builds upon names return values, do you mind if we punt on this in favor of fixing after the current PRs are merged?

Reply:
It makes me a little nervous, I've merged a lot of code planning to clean it up eventually and then forgotten to. If you feel like it would derail your progress though I'm ok with punting, this is definitely something that should get cleaned up though.